test/infer/test_gradient.py

```diff
@@ -133,3 +133,216 @@ def actual_loss_fn(params_raw):
         logger.info("actual   {} = {}".format(name, actual_grads[name]))
 
     assert_equal(actual_grads, expected_grads, prec=0.02)
+
+
+def kl_model_0_z1z2z3(params):
+    # Model
+    # z1
+    #
+    # z2
+    #
+    # z3
+    probs_z1 = jnp.array([0.5, 0.5])
+    probs_z2 = jnp.array([0.7, 0.3])
+    probs_z3 = jnp.array([0.8, 0.2])
+    pyro.sample("z1", dist.Categorical(probs_z1))
+    pyro.sample("z2", dist.Categorical(probs_z2))
+    pyro.sample("z3", dist.Categorical(probs_z3))
+
+
+def kl_model_1_z1z2z3(params):
+    # Model
+    # z1 --+
+    # |    |
+    # v    |
+    # z2   |
+    # |    |
+    # v    |
+    # z3 <-+
+    probs_z1 = jnp.array([0.5, 0.5])
+    probs_z2 = jnp.array([[0.4, 0.6], [0.6, 0.4]])
+    probs_z3 = jnp.array([[[0.4, 0.6], [0.1, 0.9]], [[0.7, 0.3], [0.2, 0.8]]])
+    z1 = pyro.sample("z1", dist.Categorical(probs_z1))
+    z2 = pyro.sample("z2", dist.Categorical(probs_z2[z1]))
+    pyro.sample("z3", dist.Categorical(probs_z3[z1, z2]))
+
+
+def kl_model_2_z2z3(params):
+    # Model (inverted the order of z1 and z2)
+    # z2 --+
+    # |    |
+    # v    |
+    # z1   |
+    # |    |
+    # v    |
+    # z3 <-+
+    probs_z2 = jnp.array([0.5, 0.5])
+    probs_z1 = jnp.array([[0.4, 0.6], [0.6, 0.4]])
+    probs_z3 = jnp.array([[[0.4, 0.6], [0.1, 0.9]], [[0.7, 0.3], [0.2, 0.8]]])
+    z2 = pyro.sample("z2", dist.Categorical(probs_z2))
+    z1 = pyro.sample("z1", dist.Categorical(probs_z1[z2]))
+    pyro.sample("z3", dist.Categorical(probs_z3[z1, z2]))
+
+
+@config_enumerate
+def kl_model_3_z2z3(params):
```

Comment:
For ease of reading the code the suffix of model names has latent sites to which analytic KL can be applied.